Batching put support using w1c workers

macintux opened the pull request:
Per discussions over the last couple of weeks, performance on large batch puts is dramatically better if those batches are parallelized in Riak. @jonmeredith pointed out that the obvious way to do so without introducing more infrastructure is to leverage the write-once workers already present on the coordinating node, hence this PR.

This changes **all** puts handled by the write-once logic, so it will need to be reviewed carefully.

Before this is approved I/we need to come up with a test to verify the behavior on worker failure and timeouts.

macintux committed:
Introduce a new asynchronous put API to allow subsystems like
time series to leverage the write-once pool as a batch processing
mechanism.

macintux committed:
Leverage the new batch processing mechanism in w1c for parallel put.

macintux committed:
Incorporate feedback:
* Configurable final cleanup delay
* Be paranoid with os:timestamp
* Record timeouts in put response

macintux committed:
Merge branch 'end-to-end/timeseries' of github.com:basho/riak_kv into feature/jrd/ts-w1c-pput

macintux committed:
Original effective timeout logic was inverted, and after further testing I determined the difference between the original timeout and the effective timeout was so small the additional complexity was of no value, so removed

macintux commented:
I performed manual testing by inserting crashes and long sleeps into the write once worker; the flow seems correct (after fixing and then removing the effective timeout logic).

macintux committed:
Failed to expurgate all StartTime references with last commit

lordnull commented on src/riak_kv_w1c_worker.erl:
```diff
@@ -354,3 +349,82 @@ erase_request_record(ReqId, #state{entries=Entries} = State) ->
+    Responses;
+async_put_reply_loop(ReqIds, Responses, Timeout) ->
+    receive
+        {'DOWN', ReqId, process, _Pid, _Reason} when is_reference(ReqId) ->
```
This concerns me a bit. If the calling process is monitoring anything other than the put worker, and that other goes down after the async_put_reply_loop starts (or even before it's able to receive that down message), it gets swallowed here, enduing up with an incorrect 'w1c_server crashed' error. This is repeated in the aysnc_put_cleanup function.

macintux replied:
Notice that it's pattern matching on the reference given back to us by monitor. Does that alleviate your concerns?

Sent from my iPhone

> On Nov 4, 2015, at 2:24 PM, Micah notifications@github.com wrote:
> 
> In src/riak_kv_w1c_worker.erl:
> 
> > -                {'DOWN', ReqId, process, _Pid, _Reason} ->
> > -                    {error, riak_kv_w1c_server_crashed};
> > -                {ReqId, Response} ->
> > -                    erlang:demonitor(ReqId, [flush]),
> > -                    Response
> > -            end
> > -    end;
> >   +synchronize_put(Error, _Options) ->
> > -    Error.
> >   +
> >   +%% Invoked by async_put_reply/2 to wait for all responses
> >   +async_put_reply_loop([], Responses, _Timeout) ->
> > -    Responses;
> >   +async_put_reply_loop(ReqIds, Responses, Timeout) ->
> > -    receive
> > -        {'DOWN', ReqId, process, _Pid, _Reason} when is_reference(ReqId) ->
> >   This concerns me a bit. If the calling process is monitoring anything other than the put worker, and that other goes down after the async_put_reply_loop starts (or even before it's able to receive that down message), it gets swallowed here, enduing up with an incorrect 'w1c_server crashed' error. This is repeated in the aysnc_put_cleanup function.
> 
> —
> Reply to this email directly or view it on GitHub.

lordnull replied:
Except it doesn't seem to here (It does in the async_put function, so no worries there). We have a list of ReqIds in the function head, the receive clause just checks that ReqId is a reference (which all monitor's are), and blindly deletes from the ReqIds list, which will work even if the key doesn't exist. Unless I've completely mis-read the code...

macintux replied:
Oh, bloody ‘ell, you’re right. This was easier when we were expecting one specific ID.

Good catch, thanks.

-John

> On Nov 4, 2015, at 3:15 PM, Micah notifications@github.com wrote:
> 
> In src/riak_kv_w1c_worker.erl https://github.com/basho/riak_kv/pull/1254#discussion_r43933086:
> 
> > -                {'DOWN', ReqId, process, _Pid, _Reason} ->
> > -                    {error, riak_kv_w1c_server_crashed};
> > -                {ReqId, Response} ->
> > -                    erlang:demonitor(ReqId, [flush]),
> > -                    Response
> > -            end
> > -    end;
> >   +synchronize_put(Error, _Options) ->
> > -    Error.
> >   +
> >   +%% Invoked by async_put_reply/2 to wait for all responses
> >   +async_put_reply_loop([], Responses, _Timeout) ->
> > -    Responses;
> >   +async_put_reply_loop(ReqIds, Responses, Timeout) ->
> > -    receive
> > -        {'DOWN', ReqId, process, _Pid, _Reason} when is_reference(ReqId) ->
> >   Except it doesn't seem to here (It does in the async_put function, so no worries there). We have a list of ReqIds in the function head, the receive clause just checks that ReqId is a reference (which all monitor's are), and blindly deletes from the ReqIds list, which will work even if the key doesn't exist. Unless I've completely mis-read the code...
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/basho/riak_kv/pull/1254/files#r43933086.

macintux committed:
* Address Micah's concerns about matching unrelated DOWN messages
  by iterating over the list of request IDs while waiting for replies
* Consolidate receive blocks between the two put code paths
* Re-introduce the timeout recalculation now that we are potentially blocking
  for long periods of time, but with a much more readable structure
* Rename bindings of request IDs to request ID tuples where appropriate
  to be explicit about their structure

macintux commented:
Unclear why dialyzer is griping. Even a local test reveals nothing other than an error exists somewhere.

lordnull commented:
+1 526880d

Ran this through some of the r_t ts's, and they pass.

Local dialyzer passes here as well; looks like bors is running a second one for whatever reason? Weird, but seems unrelated to the change. The tests that are failing appear to be query parsing and compilation, something this does't touch at all.

macintux commented:
Unit tests pass locally, dialyzer still failing locally with no useful information even with `--verbose` added.

macintux committed:
Dialyzer fix, courtesy Micah

macintux commented:
Introduced some sleeps and crashes to the worker processes, things behave as expected.

Pushed a dialyzer fix, but the unit tests are probably doomed to fail until all of the renaming PRs are merged (at which point this branch might need rebasing, but I doubt it).

macintux commented:
Ah, phooey. @lordnull I can't retry the build until you +1 the latest commit

hazen commented:
@macintux Hope this helps

hazen commented:
:+1:  e0af5b1

macintux commented:
@borshop merge
